### PR TITLE
feat(medusa, workflows): refactor `generate` line items

### PR DESCRIPTION
**Pull request by fPolic**
**What**
- Create a new script that allow us to populate the product especially when we are working in isolation
- fix addProducts to shipping profile when isolation flag is enabled
- When attaching inventory items to variant, there shouldn't be a constraint check in isolation (for now?)
- fix product module unique constraints on deleted entities
- fix create cart workflow sales channels management, guarded by flag
- add scripts to assign product to profile
- fix cart update line item confirm inventory
- fix shipping retrieval from cart
- fix shipping profile service get map and the impact
- fix cart completion
- fix reservation variant retrieval
- fix reservation when variant does not manager inventory and there is no inventory module
- refactor `LineItems.generate` to not use product service
- update the create cart workflow

**Flow to follow to tests the checkout flow in an isolated mode**

- set your flags such as workflows is true, sales channels is false, isolated product domain is true
```js
featureFlags: {
    sales_channels: false,
    isolate_product_domain: true,
    workflows: true,
  },
```
- Config the product module
```js
modules: {
    // ...
    productService: {
      scope: "internal",
      resources: "isolated",
      resolve: "@medusajs/product",
      options: {
        database: {
          clientUrl: "postgresql://localhost:5432/medusa-store"
        }
      }
    },
  }
```
- update the env file 
```bash
PRODUCT_POSTGRES_URL=<your url>
```
- run the migrations
```bash
medusa migrations run
```
- run the product module migration
```bash
node node_modules/@medusajs/product/dist/scripts/bin/run-migration-up.js
```
- run the seed
```bash
yarn seed
```
- run the pricing module seed

```bash
 node node_modules/@medusajs/medusa/dist/scripts/migrate-to-pricing-module.js
```

- run the product isolated seed
```bash
node node_modules/@medusajs/medusa/dist/scripts/seed-isolated-products.js
```
- run the product shipping profile isolated seed
```bash
node node_modules/@medusajs/medusa/dist/scripts/add-product-to-shipping-profiles-isolated.js
```

**NOTE**
The only flow that does not work is without the inventory module but with manage_inventory.

### CODE REVIEW

**Review by adrien2p — COMMENTED** [review]
Dont forget to add that if the isolated flag is enabled we need to use the remote query to get the variants instead of the productVariantService ;)

File: packages/medusa/src/api/routes/store/carts/create-cart.ts
Comment (adrien2p) [inline packages/medusa/src/api/routes/store/carts/create-cart.ts]: todo: rm the todo

File: packages/medusa/src/api/routes/admin/draft-orders/create-line-item.ts
Comment (adrien2p) [inline packages/medusa/src/api/routes/admin/draft-orders/create-line-item.ts]: todo: wrong default used here

File: packages/medusa/src/api/routes/store/carts/__tests__/create-cart.js
Comment (adrien2p) [inline packages/medusa/src/api/routes/store/carts/__tests__/create-cart.js]: suggestion: store the variantId is a const and use it below

File: packages/medusa/src/api/routes/store/carts/__tests__/create-line-item.js
Comment (adrien2p) [inline packages/medusa/src/api/routes/store/carts/__tests__/create-line-item.js]: todo: this file, you have removed the fail test case, I suggest that we keep it

File: packages/medusa/src/api/routes/store/carts/create-cart.ts
Comment (adrien2p) [inline packages/medusa/src/api/routes/store/carts/create-cart.ts]: todo: why? here we just need some product/variant prop but not all, I suggest to have something related to this endpoint and the needs instead of fetching the entire relations tree. wdyt?

File: packages/medusa/src/api/routes/store/carts/create-line-item/utils/handler-steps.ts
Comment (adrien2p) [inline packages/medusa/src/api/routes/store/carts/create-line-item/utils/handler-steps.ts]: todo: same as my previous comment

File: packages/medusa/src/services/__tests__/draft-order.js
Comment (adrien2p) [inline packages/medusa/src/services/__tests__/draft-order.js]: todo: this is a breaking change, should we keep the camelcase?

Comment (adrien2p) [reply packages/medusa/src/services/__tests__/draft-order.js]: if so then I let you check elsewhere

File: packages/medusa/src/services/draft-order.ts
Comment (adrien2p) [inline packages/medusa/src/services/draft-order.ts]: todo: to prevent reducing the performance, I would suggest to list the variant before that are in the ids of the items.variant_id and create a map of it. Then you can just do the usual map to prepare and generate

File: packages/medusa/src/services/line-item.ts
Comment (adrien2p) [inline packages/medusa/src/services/line-item.ts]: todo: This is a complete breaking changes of the existing method. Should we revert and just accept variantId or data accepting another type and then handle it? eventually we can create override declaration an mark all the previous api as deprecated to be able to remove them at somepoint.

cc @olivermrbl

File: packages/types/src/line-item/common.ts
Comment (adrien2p) [inline packages/types/src/line-item/common.ts:7]: todo: since all the DTO are part of different domain and the domain can be replaced, maybe we can add the relations as unknown in a DTO. We can then type it when we consume it with the type we expect it to be. Maybe something to think about in general when we are referencing relation across modules.

cc @carlos-r-l-rodrigues @olivermrbl

File: packages/utils/src/line-items/prepare-line-item-data.ts
Comment (adrien2p) [inline packages/utils/src/line-items/prepare-line-item-data.ts]: suggestion: Generate the expected shaped data to generate a line item from the variant data

File: packages/utils/src/line-items/prepare-line-item-data.ts
Comment (adrien2p) [inline packages/utils/src/line-items/prepare-line-item-data.ts:40]: todo: I would expect an object argument to make it more explicit what is expected instead of calling `fn(data, 1, 2)` it would be `fn({ variantData, quantify: 1, unit_price: 2})` this is for a comprehensive API of the functions we are building in those cases. wdyt?

File: packages/utils/src/line-items/validate-items-input.ts
Comment (adrien2p) [inline packages/utils/src/line-items/validate-items-input.ts]: todo: Partial is not necessary here since variant_id is already optional

File: packages/workflows/src/handlers/cart/generate-line-items.ts
Comment (adrien2p) [inline packages/workflows/src/handlers/cart/generate-line-items.ts]: todo: rm

File: packages/workflows/src/handlers/cart/prepare-line-items-generation.ts
Comment (adrien2p) [inline packages/workflows/src/handlers/cart/prepare-line-items-generation.ts]: suggestion: cart

File: packages/workflows/src/handlers/cart/prepare-line-items-generation.ts
Comment (adrien2p) [inline packages/workflows/src/handlers/cart/prepare-line-items-generation.ts]: suggestion: cart

**Review by adrien2p — CHANGES_REQUESTED** [review]

File: packages/medusa/src/api/routes/store/carts/create-cart.ts
Comment (fPolic) [reply packages/medusa/src/api/routes/store/carts/create-cart.ts]: Agree, changed to fetch only necessary things

File: packages/utils/src/line-items/prepare-line-item-data.ts
Comment (fPolic) [reply packages/utils/src/line-items/prepare-line-item-data.ts:40]: I think that is more or less the same if we have less than 5 args. Not sure that this is less comprehensive since we have argument names and types.

Comment (adrien2p) [reply packages/utils/src/line-items/prepare-line-item-data.ts:40]: let's keep it as is for now

File: packages/medusa/src/services/__tests__/draft-order.js
Comment (fPolic) [reply packages/medusa/src/services/__tests__/draft-order.js]: Since we are introducing a breaking change in the `lineItemsService.generate` API anyway, with this change we can pass that argument object to repo services that need `variant_id` without having to rename it

Comment (adrien2p) [reply packages/medusa/src/services/__tests__/draft-order.js]: but we shouldn't break the API of the generate method either

Comment (fPolic) [reply packages/medusa/src/services/__tests__/draft-order.js]: does it make sense to then create a separate method in the line item service like: `generateWithIsolatedProductModule` and call this method when the flag is on?

File: packages/medusa/src/services/line-item.ts
Comment (fPolic) [reply packages/medusa/src/services/line-item.ts]: I've updated the service to not have breaking changes as we discussed

File: packages/medusa/src/api/routes/admin/draft-orders/create-line-item.ts
Comment (olivermrbl) [inline packages/medusa/src/api/routes/admin/draft-orders/create-line-item.ts:135]: todo: Throw if variant doesn't exist

File: packages/medusa/src/api/routes/store/carts/create-cart.ts
Comment (olivermrbl) [inline packages/medusa/src/api/routes/store/carts/create-cart.ts:233]: question: What happens if no variants with provided IDs exist and are passed here?

File: packages/medusa/src/api/routes/store/carts/create-line-item/utils/handler-steps.ts
Comment (olivermrbl) [inline packages/medusa/src/api/routes/store/carts/create-line-item/utils/handler-steps.ts:47]: todo: Throw if variant doesn't exist

File: packages/medusa/src/services/line-item.ts
Comment (olivermrbl) [inline packages/medusa/src/services/line-item.ts]: todo:

Comment (olivermrbl) [reply packages/medusa/src/services/line-item.ts]: Can we add this now, or will it be included in a follow up?

Comment (fPolic) [reply packages/medusa/src/services/line-item.ts]: Will add it now

File: packages/medusa/src/api/routes/store/carts/create-cart.ts
Comment (fPolic) [reply packages/medusa/src/api/routes/store/carts/create-cart.ts:233]: ~We get an empty array. I created a validation util for this that is used in the workflow [here](https://github.com/medusajs/medusa/pull/5148/files#diff-b674642bc6f1c34458a9e86095a82d1424f825ec84f6f057fc2d6f4879e52e44R47). We should add this here as well, wdyt?~
Edit: I've added `validateItemsInput` call after variants are fetched

**Review by olivermrbl — APPROVED** [review]
LGTM - pending @adrien2p's approval

File: packages/medusa/src/services/line-item.ts
Comment (olivermrbl) [inline packages/medusa/src/services/line-item.ts]: question: Just stumbled into this one. This will be a blocker no? I believe we are retrieving variants as part of collecting pricing for variants.

Comment (adrien2p) [reply packages/medusa/src/services/line-item.ts]: I don't think this will be a blocker, but maybe the function should accept the variant_id or the variant as part of the input data? like we do in some places to not have to refectch the data that we already have. Also, when calculating the prices it makes more sense that it only takes care of the calculation based on the input data rather than fetching it. wdyt?

File: packages/medusa/src/api/routes/store/carts/create-cart.ts
Comment (adrien2p) [inline packages/medusa/src/api/routes/store/carts/create-cart.ts]: todo: I believe it is handle as part of `validateItemsInput`, should we remove it?

File: packages/medusa/src/api/routes/store/carts/create-line-item/utils/handler-steps.ts
Comment (adrien2p) [inline packages/medusa/src/api/routes/store/carts/create-line-item/utils/handler-steps.ts]: todo: rm for the same reason

File: packages/medusa/src/api/routes/admin/draft-orders/create-line-item.ts
Comment (adrien2p) [inline packages/medusa/src/api/routes/admin/draft-orders/create-line-item.ts]: todo: I believe it is handled by the utils below, should we remove that?

File: packages/medusa/src/services/line-item.ts
Comment (adrien2p) [reply packages/medusa/src/services/line-item.ts]: but in any case, there is a bit more work to do here @fPolic

File: packages/medusa/src/services/line-item.ts
Comment (adrien2p) [inline packages/medusa/src/services/line-item.ts]: suggestion: rm, this method can be used no matter if the flag is enabled or not in fact

File: packages/medusa/src/utils/line-items/retrieve-variants-woth-isolated-product-module.ts
Comment (adrien2p) [inline packages/medusa/src/utils/line-items/retrieve-variants-woth-isolated-product-module.ts]: suggestion(non blocking): Are we sure all the places where this util is used require all the same fields? I believe this function can take `fields` as an arguments which would looks like `["id", "title", ..., "variants.id", ...]` and you can here generate the remote query object using the util we talked about.

File: packages/utils/src/line-items/validate-items-input.ts
Comment (adrien2p) [inline packages/utils/src/line-items/validate-items-input.ts:10]: suggestion: Maybe rename `validateLineItemsVariantsIsPresent`

**Review by adrien2p — CHANGES_REQUESTED** [review]
Almost LG, few more comments and one place to add some work. Around the pricing service. You can create an overload and allow the pass the variant data instead of just the id, and in that case if the data are prensent then we don't fetch the variants inside the pricing service method

File: packages/medusa/src/services/line-item.ts
Comment (fPolic) [reply packages/medusa/src/services/line-item.ts]: On it :)

File: packages/medusa/src/api/routes/admin/draft-orders/create-line-item.ts
Comment (fPolic) [reply packages/medusa/src/api/routes/admin/draft-orders/create-line-item.ts]: yep handled in the util as well...removed

File: packages/medusa/src/services/line-item.ts
Comment (fPolic) [reply packages/medusa/src/services/line-item.ts]: true, meant to emphasize that it is now only used for IsolatedProductModule purposes

**Comment by olivermrbl** [comment]
In the call `ProductVariantInventoryService.confirmInventory`, we are still querying for variants, which is currently blocking me from adding to cart. I believe this is the last place that needs refactoring.

File: packages/utils/src/index.ts
Comment (carlos-r-l-rodrigues) [inline packages/utils/src/index.ts:6]: Should all these utils function be in the utils package?
aren't they specific to the /medusa package?

**Review by carlos-r-l-rodrigues — APPROVED** [review]

File: packages/utils/src/index.ts
Comment (adrien2p) [reply packages/utils/src/index.ts:6]: They can be useful in the workflows as well, and they are standalone, so i would keep there in utils for now. Wdyt?

Comment (fPolic) [reply packages/utils/src/index.ts:6]: Yes, as Adrien said it they are used in workflows as well

File: packages/medusa/src/repositories/shipping-profile.ts
Comment (olivermrbl) [inline packages/medusa/src/repositories/shipping-profile.ts]: todo: Change to follow same function definition format as other repositories
```suggestion
    async addProducts(
```

File: packages/medusa/src/scripts/seed-isolated-products.ts
Comment (olivermrbl) [inline packages/medusa/src/scripts/seed-isolated-products.ts]: question: Should we delete this now that we have confirmed the flow works?

File: packages/medusa/src/services/cart.ts
Comment (olivermrbl) [inline packages/medusa/src/services/cart.ts]: question: Should we have a type for this? I imagine we'll use this more going forward

File: packages/medusa/src/services/line-item.ts
Comment (olivermrbl) [inline packages/medusa/src/services/line-item.ts:430]: comment: Breaking change

File: packages/medusa/src/strategies/cart-completion.ts
Comment (olivermrbl) [inline packages/medusa/src/strategies/cart-completion.ts]: question: Can we type this more appropriately?

**Review by olivermrbl — CHANGES_REQUESTED** [review]
Great piece of work. Only had a couple of minor comments to address before we merge 💪

File: packages/medusa/src/services/cart.ts
Comment (fPolic) [reply packages/medusa/src/services/cart.ts]: I've added `RemoteQueryQuery` type to `types`

File: packages/medusa/src/services/line-item.ts
Comment (fPolic) [reply packages/medusa/src/services/line-item.ts:430]: so we should use another name for the new method and keep this one?

**Comment by adrien2p** [comment]
should we merge [that ](https://github.com/medusajs/medusa/pull/5701)first and then update this pr or the other way around?

**Comment by fPolic** [comment]
> should we merge [that ](https://github.com/medusajs/medusa/pull/5701)first and then update this pr or the other way around?

think this is ready for another round of review since it was a bunch of merge conflicts and we should be good to merge then

**Pull request closed without merging**
